adrs/0049-prebuilt-rubies-at-runtime.md

```diff
@@ -3,43 +3,78 @@
 ***Status***: proposed
 
 ## Context
-Currently, only two versions of Ruby are available at runtime as part of the hosted build VM tools cache.  The setup-ruby action only supports using what's in the cache and does not support pulling other versions at runtime.  It also doesn't support pulling other versions like JRuby at runtime.
+Currently, only a few versions of Ruby are available at runtime prebuilt by the [virtual-environment images](https://github.com/actions/virtual-environments) as part of the hosted build VM tools cache.  The setup-ruby action only supports using what's in the cache and does not support acquiring other versions at runtime. 
+
+This introduces two problems:
+
+  1.  New version latency: When new versions are released, they are not available for use until a new image is released.
+  2.  Versions can be removed.  The current image policy is latest of each minor version.  Even if that is increased to n - 1, [builds that were working can break](https://github.com/actions/virtual-environments/issues/281).  And since images roll through scale units with the readme only being updated once it's everywhere, it's hard to plan for the change and react.
 
 This is in contrast to other setup actions like setup-node where the cache is an optimization but supports pulling any version by a specified semver at runtime.
 
+Ruby officially only supports source code distribution.  There are no "official" pre built distributions since there are system and environmental dependencies.
+
 ## Goals
 
 - Offer as many versions as possible by platforms
 - Offer popular Ruby variants like JRuby
 - Consistent with other setup-xxx actions patterns and Actions workflows.
+- Ideally works across all supported scenarios including self-hosted runners.
+- Ideally works on all the supported runner platforms (windows, mac-os, all linux variants)
+- Ideally works with actions container scenarios
 
-## Decisions
+## Options
 
-### Prebuilt Rubies
+### Download Prebuilt Rubies on Misses
 
 Other tools like offer a [distribution](https://nodejs.org/dist/) which offers a [queryable endpoint](https://nodejs.org/dist/index.json) which allows the desired version [by a version spec which is semver](https://github.com/actions/setup-node#setup-node).
 
 Other CI services like Travis support pulling a wide variety of [ruby versions offered here](http://rubies.travis-ci.org/).
 
-Requirements:
-- Offer prebuilt Ruby, JRuby, TruffleRuby versions 
```

Comment:
Could you add TruffleRuby in the `Goals` above since it's removed here?

Reply:
I'm focusing this change request on (1) getting versions of Ruby at runtime and (2) self-hosted scenarios for on-prem and GHES scenarios (hi pri right now).  We can write another change request on adding variations.  We still haven't done a good job with Ruby ...